### ISSUE TITLE
With auto, only enable compression if lz4 achieves a certain percentage reduction?

### ISSUE THREAD

**Issue by jdchristensen** [borgbackup/borg#2339]
With `auto,C`, the current code runs the compressor C if lz4 manages to reduce the size of the chunk at all.  In many cases, lz4 does reduce the chunk size, but by less than 1%.  Here are the times for backing up some jpg files (unencrypted, local SSD to itself) with various compression options.  The last one changes the line
```
        if cdata_len < data_len:
```
to
```
        if cdata_len < 0.95*data_len:
```
so that the compressor is only used if lz4 gives at least a 5% reduction.

```
  none       94.12MB  1.2s
  lz4        93.96MB  1.4s
  zlib       92.89MB  4.3s
  auto,zlib  93.24MB  2.7s
  5%         94.08MB  1.4s
```
The time savings is significant, bringing it down to the time needed for just lz4, with little loss in compression.  Does this seem reasonable?  5%?  3%?

**Comment by enkore** [comment]
5 % sounds like a starting point. Maybe make that part of the specification, e.g `--compression auto,5,lzma,4`? (OTOH, unneeded flexibility is just useless complexity).

**Comment by ThomasWaldmann** [comment]
guess we could just hardcode some reasonable value, 0.95 or rather 0.98?

**Comment by enkore** [comment]
9.31415926535897932384626433832795028841971693993751058 ?

**Comment by jdchristensen** [comment]
People who use auto are probably doing so because they want to avoid unnecessary cpu overhead, so missing a few percent of compression is probably ok.  I did some tests on jpg files using 1, 0.99, 0.97 and 0.95, and you get most of the benefit at 0.97, with a little more at 0.95.  I think the value should just be hardcoded, since neither the time nor the compression change much once you are in the 0.95 to 0.97 range.  How about 0.97?  Should I make a pull request?

**Comment by jdchristensen** [comment]
For the jpg data I tested on, here are the compression ratios for the 115 chunks, sorted.  You can see how a cutoff in the range of .95 to .97 does a good job.

```
0.531819
0.826581
0.932987
0.959410
0.960385
0.964171
0.967969
0.969889
0.973682
0.976139
0.981434
0.981744
0.985982
0.986575
0.988951
0.989169
0.989760
0.990135
0.991613
0.992228
0.992652
0.994379
0.994408
0.994469
0.994866
0.995089
0.995232
0.995860
0.995884
0.995906
0.996716
0.996733
0.998443
0.998810
0.999396
0.999985
1.000081
1.000322
1.000409
1.000450
1.000671
1.000819
1.000867
1.000935
1.001139
1.001204
1.001470
1.001592
1.002043
1.002133
1.002218
1.002268
1.002373
1.002447
1.002447
1.002486
1.002514
1.002518
1.002535
1.002580
1.002811
1.003102
1.003189
1.003198
1.003219
1.003257
1.003284
1.003308
1.003332
1.003538
1.003624
1.003751
1.003852
1.003881
1.003916
1.003921
1.003922
1.003923
1.003923
1.003923
1.003923
1.003923
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003924
1.003925
1.003925
1.003925
1.003925
1.003926
1.003926
1.003926
1.003927
1.003927
1.003928
1.003928
1.003928
1.003928
1.003929
1.003929
1.003929
1.003931
1.003931
1.003932
1.004575
```

**Comment by enkore** [comment]
Sounds good to me

**Comment by jdchristensen** [comment]
Fixed in PR #2340
